### ISSUE TITLE
custom postscripts in subfolder does not preserve or have execute permissions

### ISSUE THREAD

**Issue by conxuro** [xcat2/xcat-core#6579]
When custom postscripts are placed in a postscript subdirectory  (e.g. /install/postscripts/IRB/) they loose the execute permissions after the installation but not if they are in the postscript default directory (e.g /install/postscripts/).

This results in permission denied errors:
```
Thu Feb 20 14:48:04 CET 2020 [info]: xcat.deployment.postscript: postscript remoteshell return with 0
Thu Feb 20 14:48:04 CET 2020 [info]: xcat.deployment.postscript: Running postscript: syncfiles
Thu Feb 20 14:48:05 CET 2020 [info]: xcat.deployment.postscript: postscript syncfiles return with 0
Thu Feb 20 14:48:05 CET 2020 [error]: xcat.deployment.postscript: postscript IRB/irb-centos_repos_disable does NOT exist.
Thu Feb 20 14:48:05 CET 2020 [info]: xcat.deployment.postscript: Running postscript: IRB/irb-network_services
/xcatpost/mypostscript: line 41: ./IRB/irb-network_services: Permission denied
Thu Feb 20 14:48:05 CET 2020 [info]: xcat.deployment.postscript: postscript IRB/irb-network_services return with 126
Thu Feb 20 14:48:05 CET 2020 [info]: xcat.deployment.postscript: Running postscript: IRB/irb-slurm_munge_useradd
/xcatpost/mypostscript: line 41: ./IRB/irb-slurm_munge_useradd: Permission denied
Thu Feb 20 14:48:05 CET 2020 [info]: xcat.deployment.postscript: postscript IRB/irb-slurm_munge_useradd return with 126
Thu Feb 20 14:48:05 CET 2020 [info]: xcat.deployment: finished firstboot preparation, sending request to 172.17.0.2:3002 for changing status...
updateflag.awk: Retrying flag update
Thu Feb 20 14:50:48 CET 2020 [info]: xcat.deployment: Running /xcatpost/mypostscript.post
Thu Feb 20 14:50:48 CET 2020 [info]: xcat.deployment.postbootscript: Running postbootscript: otherpkgs
Thu Feb 20 14:50:48 CET 2020 [info]: xcat.deployment.postbootscript: postbootscript otherpkgs return with 0
Thu Feb 20 14:50:48 CET 2020 [info]: xcat.deployment.postbootscript: Running postbootscript: confignetwork
Thu Feb 20 14:51:15 CET 2020 [info]: xcat.deployment.postbootscript: postbootscript confignetwork return with 0
Thu Feb 20 14:51:15 CET 2020 [error]: xcat.deployment.postbootscript: postbootscript IRB/irb-slurm_munge_enable_start does NOT exist.
Thu Feb 20 14:51:38 CET 2020 [error]: xcat.deployment.postbootscript: provision completed with error.(nodo20)
Thu Feb 20 14:51:38 CET 2020 [info]: xcat.deployment: /xcatpost/mypostscript.post return
Thu Feb 20 14:51:38 CET 2020 [info]: xcat.deployment: =============deployment ending====================
```
---
Software used:
* Management node
```
$ cat /etc/redhat-release 
CentOS Linux release 7.7.1908 (Core)
$ xcatconfig -v
Version 2.15 (git commit 218c6d3acc8bdbd7f72115e48cda2b1a3613d18a, built Mon Nov  4 15:17:59 EST 2019)
```

* Compute node
```
[root@nodo20 ~]# cat /etc/redhat-release 
CentOS Linux release 8.0.1905 (Core)
```

**Comment by besawn** [comment]
Can you please provide the permissions and ownership of the directory and custom postscripts from both the management node and compute node?

On management node:
```
ls -ald /install/postscripts/IRB/
ls -al /install/postscripts/IRB/
```

On compute node:
```
ls -ald /xcatpost/IRB/
ls -al /xcatpost/IRB/
```

**Comment by conxuro** [comment]
* On management node:
```
$ ls -ald /install/postscripts/IRB
drwxr-xr-x 2 root root 4096 Feb 20 14:10 /install/postscripts/IRB

$ ls -al /install/postscripts/IRB
total 28
drwxr-xr-x 2 root root 4096 Feb 20 14:10 .
drwxr-xr-x 8 root root 4096 Feb 20 14:10 ..
-rwxr-xr-x 1 root root  102 Feb 19 13:56 irb-centos_repos_disable
-rwxr-xr-x 1 root root   98 Feb 20 12:35 irb-confignetwork
-rwxr-xr-x 1 root root  758 Feb 20 14:08 irb-network_services
-rwxr-xr-x 1 root root  114 Feb 19 11:45 irb-slurm_munge_enable_start
-rwxr-xr-x 1 root root  332 Feb 19 15:40 irb-slurm_munge_useradd
```
* On compute node just after install:
```
[root@nodo20 IRB]# ls -ald /xcatpost/IRB/
drwxr-xr-x. 2 root root 4096 Feb 20 16:19 /xcatpost/IRB/

[root@nodo20 IRB]# ls -al /xcatpost/IRB/
total 20
drwxr-xr-x. 2 root root 4096 Feb 20 16:19 .
drwxr-xr-x. 8 root root 4096 Feb 20 16:19 ..
-rw-r--r--. 1 root root   98 Feb 20 16:19 irb-confignetwork
-rw-r--r--. 1 root root  758 Feb 20 16:19 irb-network_services
-rw-r--r--. 1 root root  332 Feb 20 16:19 irb-slurm_munge_useradd
```

And the permissions preserved (or with +x added after) outsite de subdirectory:

* Management node:
```
[root@hpchead slurm]# ls -l /install/postscripts/irb*
-rwxr-xr-x 1 root root 102 Feb 19 13:56 /install/postscripts/irb-centos_repos_disable
-rwxr-xr-x 1 root root  98 Feb 20 12:35 /install/postscripts/irb-confignetwork
-rwxr-xr-x 1 root root 758 Feb 20 14:08 /install/postscripts/irb-network_services
-rwxr-xr-x 1 root root 114 Feb 19 11:45 /install/postscripts/irb-slurm_munge_enable_start
-rwxr-xr-x 1 root root 332 Feb 19 15:40 /install/postscripts/irb-slurm_munge_useradd
```
* Compute node:
```
[root@nodo20 IRB]# ls -l /xcatpost/irb-*
-rwxr-xr-x. 1 root root  98 Feb 20 16:19 /xcatpost/irb-confignetwork
-rwxr-xr-x. 1 root root 758 Feb 20 16:19 /xcatpost/irb-network_services
-rwxr-xr-x. 1 root root 332 Feb 20 16:19 /xcatpost/irb-slurm_munge_useradd
```

**Comment by conxuro** [comment]
The issue seems related to:
```
/opt/xcat/share/xcat/install/scripts/post.xcat.ng:148:find /xcatpost -type f -maxdepth 1 -exec chmod 0755 {} +
```
Not sure about the best fix, because for example, -maxdepth 2 would not be a good solution because we don't want execution permissions in /xcatpost/ca and other subdirectories.

Is there any option to know the custom postscripts subdirectories? Or maybe to parse postscripts table?


Also that commands throws a warning during the installation:
```
find: warning: you have specified the -maxdepth option after a non-option argument -type, but options are not positional (-maxdepth affects tests specified before it as well as those specified after it).  Please specify options before other arguments.
```

**Comment by peterwywong** [comment]
Based on https://xcat-docs.readthedocs.io/en/stable/guides/admin-guides/manage_clusters/common/deployment/prepostscripts/post_script.html, to add your own script, place it in /install/postscripts on the management node. Make sure it is executable and world readable. 

It means any new postscripts created by users should have the 0755 permission.

Therefore, the following might work.

find /xcatpost -maxdepth 2 -type d \( -name ca -o -name hostkeys -o -name repos -o -name _ssh -o -name _xcat \) -prune -o -print -exec chmod 0755 {} +

It sets all files with the 755 permission except those under ca, hostkeys, respos, _ssh and _xcat.

**Comment by besawn** [comment]
@peterwywong This approach could work, but I would like us to do some more internal investigation. Can you please try to reproduce this issue internally using the autotest framework?

@conxuro Thank you for the detailed debug information. We will provide more information as it becomes available.

**Comment by peterwywong** [comment]
@conxuro We have got more information about this issue.

With xCAT 2.15.1, diskful and diskless installations were conducted with RHEL 7.6 on a management node and a compute node, and another round with RHEL 8.1. We have not reproduced the problem where the execute permission is not preserved.

To process postscripts, /install/postscripts/xcatdsklspost on MN is called.

On Line 111 of xcatdsklspost, wget has all files under /install/postscripts copied from MN to CN:
  export LANG=C; wget -l inf -nH -N -r --waitretry=10 --random-wait -e robots=off -T 60 -nH --cut-dirs=2 --reject "index.html*" --no-parent http://$server$INSTALLDIR/postscripts/ -P /$xcatpost 2> /tmp/wget.log

If a file has the 444 permission, it is changed to 644 by wget.

And then Line 687 adds "x" to all files recursively except for those files under a few directories:
  chmod -R +x `find $xcatpost/ -maxdepth 1 -print | grep -E -v "^($xcatpost/|$xcatpost/_xcat|$xcatpost/_ssh|$xcatpost/ca|$xcatpost/hostkeys)$"`

That is, 644 will become 755.

If the original file has 755 on MN, it should still be 755 on CN by the above codes.

We also found that the following command in /opt/xcat/share/xcat/install/scripts/post.xcat.ng does not add "+x" recursively.
  find /xcatpost -maxdepth 2 -type d \( -name ca -o -name hostkeys -o -name repos -o -name _ssh -o -name _xcat \) -prune -o -print -exec chmod 0755 {} + 

Which xCAT version do you use?  Would you provide more details about how CentOS 8.0 is installed on CN by the MN which runs CentOS 7.7? Do you run "rinstall <node definition> osimage=<CentOS 8.1> on MN?

The following might be helpful.

Could you add some echolog statement in /install/postscripts/xcatdsklspost (on MN) before Line 687 to see whether "chmod -R +x" is actually executed?

Hope that we are getting closer to the problem.

**Comment by cxhong** [comment]
this issue is fixed in the PR #6669
